File: tools/tilemap.py
#!/usr/bin/env python3
"""Losslessly convert little-endian 16-bit tilemaps to readable text."""
import argparse
import re
import struct
from pathlib import Path


WORD = re.compile(r"[0-9a-fA-F]{4}")


def export_tilemap(data, width):
    if width <= 0 or len(data) % 2:
        raise ValueError("tilemap must contain whole 16-bit entries")
    values = struct.unpack(f"<{len(data) // 2}H", data)
    if not values or len(values) % width:
        raise ValueError("width must divide the tilemap entry count")
    return "".join(
        " ".join(f"{value:04x}" for value in values[start:start + width]) + "\n"
        for start in range(0, len(values), width)
    )


def import_tilemap(text):
    tokens = text.split()
    if not tokens or any(WORD.fullmatch(token) is None for token in tokens):
        raise ValueError("tilemap text must contain four-digit hex words")
    return b"".join(struct.pack("<H", int(token, 16)) for token in tokens)


def self_test():
    raw = b"".join(struct.pack("<H", index * 257) for index in range(64))
    if import_tilemap(export_tilemap(raw, 8)) != raw:
        raise AssertionError("tilemap round-trip failed")
    print("self-test=ok")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--self-test", action="store_true")
    commands = parser.add_subparsers(dest="command")
    export = commands.add_parser("export")
    export.add_argument("input", type=Path)
    export.add_argument("--width", type=int, required=True)
    export.add_argument("-o", "--output", type=Path, required=True)
    build = commands.add_parser("import")
    build.add_argument("input", type=Path)
    build.add_argument("-o", "--output", type=Path, required=True)
    args = parser.parse_args()
    if args.self_test:
        self_test()
        if args.command is None:
            return
    if args.command == "export":
        result = export_tilemap(args.input.read_bytes(), args.width)
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_text(result)
        print(f"entries={len(result.split())}")
    elif args.command == "import":
        result = import_tilemap(args.input.read_text())
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_bytes(result)
        print(f"bytes={len(result)}")
    else:
        parser.error("a tilemap command is required")


if __name__ == "__main__":
    main()

File: tools/tilemap.ts
// Tool role: both; imported by tools/build_assets.ts, tools/chiiki_map_resources.ts, tools/map_container_components.ts (+1 more); invoked by package.json.
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

const WORD = /^[0-9a-fA-F]{4}$/;

export function export_tilemap(data: Uint8Array, width: number): string {
  if (width <= 0 || data.length % 2 !== 0)
    throw new Error("tilemap must contain whole 16-bit entries");
  const count = data.length / 2;
  if (count === 0 || count % width !== 0)
    throw new Error("width must divide the tilemap entry count");
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let result = "";
  for (let start = 0; start < count; start += width) {
    const row: string[] = [];
    for (let index = start; index < start + width; index++)
      row.push(view.getUint16(index * 2, true).toString(16).padStart(4, "0"));
    result += `${row.join(" ")}\n`;
  }
  return result;
}

export function import_tilemap(text: string): Buffer {
  const tokens = text.split(/\s+/).filter(Boolean);
  if (tokens.length === 0 || tokens.some((token) => !WORD.test(token)))
    throw new Error("tilemap text must contain four-digit hex words");
  const result = Buffer.alloc(tokens.length * 2);
  tokens.forEach((token, index) => result.writeUInt16LE(Number.parseInt(token, 16), index * 2));
  return result;
}

export function self_test(): void {
  const raw = Buffer.alloc(128);
  for (let index = 0; index < 64; index++) raw.writeUInt16LE(index * 257, index * 2);
  if (!import_tilemap(export_tilemap(raw, 8)).equals(raw))
    throw new Error("tilemap round-trip failed");
  console.log("self-test=ok");
}

function value(args: string[], name: string): string {
  const index = args.indexOf(name);
  if (index < 0 || index + 1 >= args.length) throw new Error(`${name} is required`);
  return args[index + 1];
}

function main(args: string[]): void {
  if (args.includes("-h") || args.includes("--help")) {
    console.log("usage: tilemap.ts [--self-test] {export,import} ...");
    return;
  }
  if (args.includes("--self-test")) {
    self_test();
    args = args.filter((arg) => arg !== "--self-test");
    if (args.length === 0) return;
  }
  const command = args[0];
  const input = args[1];
  const output = value(args, args.includes("-o") ? "-o" : "--output");
  mkdirSync(dirname(output), { recursive: true });
  if (command === "export") {
    const result = export_tilemap(readFileSync(input), Number.parseInt(value(args, "--width"), 10));
    writeFileSync(output, result);
    console.log(`entries=${result.trim().split(/\s+/).length}`);
  } else if (command === "import") {
    const result = import_tilemap(readFileSync(input, "utf8"));
    writeFileSync(output, result);
    console.log(`bytes=${result.length}`);
  } else throw new Error("a tilemap command is required");
}

if (import.meta.main) main(process.argv.slice(2));
